torch/csrc/jit/passes/pattern_fusion.cpp

```diff
@@ -0,0 +1,134 @@
+#include <torch/csrc/jit/passes/pattern_fusion.h>
+#include <torch/csrc/jit/irparser.h>
+#include <torch/csrc/jit/subgraph_matcher.h>
+
+namespace torch {
+namespace jit {
+
+void PatternFuser::RegisterDefaultPatterns() {
+  // TODO: Add actual patterns (like Conv-Relu).
+  RegisterFusionPattern(
+      R"IR(
+graph(%x, %w, %b):
+  %c = aten::conv(%x, %w, %b)
+  %r = aten::relu(%c)
+  return (%r))IR",
+      /*fused_node_name=*/"aten::convrelu",
+      /*inputs=*/{"x", "w", "b"},
+      /*outputs=*/{"r"});
+}
+
+void PatternFuser::RegisterFusionPattern(
+    const std::string& pattern,
+    const std::string& fused_node_name,
+    std::vector<std::string> inputs,
+    std::vector<std::string> outputs) {
+  FusionPatternDescr d = {pattern, fused_node_name, inputs, outputs};
+  patterns_.push_back(d);
+}
+
+std::shared_ptr<script::Module> PatternFuser::runOnModule(
+    std::shared_ptr<script::Module> module) {
+  nodes_to_delete_.clear();
+  const auto& methods = module->get_methods();
+  for (const auto& m : methods) {
+    auto g = m->function().graph();
+    runOnGraph(g);
+  }
+  return module;
+}
+
+void PatternFuser::runOnGraph(std::shared_ptr<Graph>& graph) {
+  for (const FusionPatternDescr& pattern : patterns_) {
+    fuseSinglePatternOnGraph(graph, pattern);
+  }
+}
+
+void PatternFuser::fuseSinglePatternOnGraph(
+    std::shared_ptr<Graph>& graph,
+    FusionPatternDescr pattern) {
+  std::unordered_map<Value*, Value*> rewrite_map;
+  std::vector<Value*> values_to_rewrite;
+
+  Graph pattern_graph;
+  std::unordered_map<std::string, Value*> vmap;
+  script::parseIR(pattern.pattern, &pattern_graph, vmap);
```

Comment:
Why do we keep reparsing the patters every time this is run? Wouldn't it be better to just do it once when it is registered?